test/FunctionalTests/Tests/DataServices/UnitTests/Client.TDD.Tests/Tests/DeepInsertSaveResultTests.cs

```diff
@@ -0,0 +1,370 @@
+            this.context.EntityTracker.DetachResource(personDescriptor);
+
+            Action action = () =>
+            {
+                this.GetBulkUpdateGraph(person);
+            };
+
+            action.ShouldThrow<InvalidOperationException>(Strings.Context_EntityNotContained);
+        }
+
+        [Fact]
+        public void SerializeEntry_With_UpdateObject_Fails()
+        {
+            var person = new Person
+            {
+                ID = 100,
+                Name = "Bing",
+            };
+
+            var cars = new List<Car>
+            {
+                new Car{
+                    ID = 200,
+                    Name = "Car 1",
+                }
+            };
+
+            this.context.AttachTo("Persons", person);
+            this.context.AttachTo("Cars", cars[0]);
+
+            DataServiceCollection<Car> carsCollection = new DataServiceCollection<Car>(this.context);
+            carsCollection.Load(cars);
+            carsCollection[0].Name = "Updated car";
+
+            this.context.AddLink(person, "Cars", carsCollection[0]);
+
+            Action action = () =>
+            {
+                this.GetBulkUpdateGraph(person);
+            };
+
+            action.ShouldThrow<InvalidOperationException>(Strings.Context_DeepInsertDeletedOrModified);
+        }
+
+        [Fact]
+        public void Serialize_MultipleTopLevel_Resources_Fails()
+        {
+            var person = new Person
+            {
+                ID = 100,
+                Name = "Bing",
+            };
+
+            var person2 = new Person
+            {
+                ID = 200,
+                Name = "Edge"
+            };
+
+            this.context.AttachTo("Persons", person);
+            this.context.AttachTo("Persons", person2);
+
+            Action action = () =>
+            {
+                this.GetBulkUpdateGraph(person, person2);
+            };
+
+            action.ShouldThrow<InvalidOperationException>(Strings.Context_DeepInsertOneTopLevelEntity);
+        }
+
+        private BulkUpdateGraph GetBulkUpdateGraph<T>(params T[] objects)
+        {
+            List<Descriptor> changedEntries = this.context.EntityTracker.Entities.Cast<Descriptor>()
+                                  .Union(context.EntityTracker.Links.Cast<Descriptor>())
+                                  .Union(context.EntityTracker.Entities.SelectMany(e => e.StreamDescriptors).Cast<Descriptor>())
+                                  .Where(o => o.IsModified && o.ChangeOrder != UInt32.MaxValue)
+                                  .OrderBy(o => o.ChangeOrder)
+                                  .ToList();
+
+            var result = new DeepInsertSaveResult(this.context, Util.SaveChangesMethodName, SaveChangesOptions.DeepInsert, null, null);
+            result.BuildDescriptorGraph<T>(changedEntries, true, (T[])objects);
+            this.bulkUpdateGraph = result.BulkUpdateGraph;
+
+            return this.bulkUpdateGraph;
+        }
+

```

Comment:
@KenitoInc Can you add a test where the nested navigation properties are single-valued?

Reply:
`SerializeEntry_With_SingleValueNavigation`
`SerializeEntry_With_SingleValueNavigationLink`